Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device.h

```diff
@@ -41,22 +40,89 @@ enum IOSReturnCode : s32
   IPC_EESEXHAUSTED = -1016,  // Max of 2 ES handles exceeded
 };
 
-// A struct for IOS ioctlv calls
-struct SIOCtlVBuffer
+struct IOSRequest
 {
-  explicit SIOCtlVBuffer(u32 address);
-
-  const u32 m_Address;
-  u32 Parameter;
-  u32 NumberInBuffer;
-  u32 NumberPayloadBuffer;
-  u32 BufferVector;
-  struct SBuffer
+  u32 address = 0;
+  IPCCommandType command = IPC_CMD_OPEN;
+  u32 fd = 0;
+  explicit IOSRequest(u32 address);
+  virtual ~IOSRequest() = default;
+  void SetReturnValue(s32 new_return_value) const;
+};
+
+enum IOSOpenMode : s32
+{
+  IOS_OPEN_READ = 1,
+  IOS_OPEN_WRITE = 2,
+  IOS_OPEN_RW = (IOS_OPEN_READ | IOS_OPEN_WRITE)
+};
+
+struct IOSOpenRequest final : IOSRequest
+{
+  std::string path;
+  IOSOpenMode flags = IOS_OPEN_READ;
+  explicit IOSOpenRequest(u32 address);
+};
+
+struct IOSReadWriteRequest final : IOSRequest
+{
+  u32 buffer = 0;
+  u32 size = 0;
+  explicit IOSReadWriteRequest(u32 address);
+};
+
+struct IOSSeekRequest final : IOSRequest
+{
+  enum SeekMode
   {
-    u32 m_Address, m_Size;
+    IOS_SEEK_SET = 0,
+    IOS_SEEK_CUR = 1,
+    IOS_SEEK_END = 2,
   };
-  std::vector<SBuffer> InBuffer;
-  std::vector<SBuffer> PayloadBuffer;
+  u32 offset = 0;
+  SeekMode mode = IOS_SEEK_SET;
+  explicit IOSSeekRequest(u32 address);
+};
+
+struct IOSIOCtlRequest final : IOSRequest
+{
+  u32 request = 0;
+  u32 buffer_in = 0;
+  u32 buffer_in_size = 0;
+  // Contrary to the name, the output buffer can also be used for input.
+  u32 buffer_out = 0;
+  u32 buffer_out_size = 0;
+  explicit IOSIOCtlRequest(u32 address);
+  void Log(const std::string& description, LogTypes::LOG_TYPE type = LogTypes::WII_IPC_HLE,
+           LogTypes::LOG_LEVELS level = LogTypes::LINFO) const;
+  void Dump(const std::string& description, LogTypes::LOG_TYPE type = LogTypes::WII_IPC_HLE,
+            LogTypes::LOG_LEVELS level = LogTypes::LINFO) const;
+  void DumpUnknown(const std::string& description, LogTypes::LOG_TYPE type = LogTypes::WII_IPC_HLE,
+                   LogTypes::LOG_LEVELS level = LogTypes::LERROR) const;
+};
+
+struct IOSIOCtlVRequest final : IOSRequest
+{
+  struct IOVector
+  {
+    u32 addr = 0;
```

Comment:
This one is a feel-free-to-ignore-me nitpick: you used `address` everywhere else but only abbreviated it here :P